### DISCUSSION TITLE
Choose a deployment platform for portfolio sites

### DISCUSSION THREAD

**Discussion by illicitonion** [CodeYourFuture/curriculum#434] [General]
We have ~hundreds of trainees and graduates who typically have 2-10 projects in a portfolio.

Portfolio sites typically have three components:
1. Frontend (often React, generally deployable as a bundle of static resources, but where forwarding all paths to a single file handler can be useful for using React Router)
2. Backend (generally Express, listening on a single port, with access to some secrets configured as env vars)
3. Database (generally Postgres)

Each of these typically has _very_ low resource requirements (Fontend consumes <50MB disk, backend consumes <100MB RAM, database stores <10MB of data) and low access rates (averaging <1 request per day, with likely peaks around 50 requests across a few minutes).

Key requirements are:
1. Free to trainees/graduates - our trainees must not need to pay to host their portfolios.
2. Low latency - the point of hosting these portfolios is for employers to be able to view and interact with them. If a website takes several seconds to start working, it reflects poorly on the trainees, and and employer is likely to give up on the site.

Ideally our trainees could host a number of these portfolio sites independently (e.g. create independent databases for each, having different DNS names or at least ports for each backend), but we've made sharing work when needed (e.g. independent tables in the same database).

Other useful features:
* Auto-deploy on git push
* Access to backend logs
* Deploy previews on pull requests

Ideally there would be a hosted service we could use for this, but if not, we could potentially spec out running something ourselves (ideally something out of the box rather than something we'd need to develop).

In the past we have used:

For frontends:
* Netlify - generally works well
* Render - generally works well

For backends:
* Render - but they aggressively spin down backends, and are _very_ slow to spin back up

For databases:
* Render - but they delete databases after 90 days
* ElephantSQL - recently stopped offering a free tier

**Comment by sztupy** [comment]
Other assessed options (see Slack):

Frontend:
* CloudFlare - generally works well
* Fly.io - frontend supported through monorepos, only a single project in the free tier

Backend:
* Netlify - needs serverless lambda wrapper for express.js, no in-built database
* Cloudflare - only serverless lambda implementations supported, not express.js
* Fly.io - only allows a single project in the free tier (or two if no database is needed)

Database:
* Cloudflare - Uses SQLite and not Postgres syntax and connectivity library
* Fly.io - PSQL, Only a single database in the free tier (or two if no backend is needed)
* Supabase - PSQL, Shuts down after one week of inactivity
